lsp/nls/src/cache.rs

```diff
@@ -73,7 +74,18 @@ impl CacheExt for Cache {
             let (_, linearized) =
                 typecheck::type_check_linearize(term, initial_ctxt.clone(), self, host, building)
                     .map_err(|err| vec![Error::TypecheckError(err)])?;
-            lin_registry.insert(file_id, linearized, term);
+
+            // Urg. We're linearizing twice...
+            let (_, type_lookups) = typecheck::type_check_linearize(
+                term,
+                initial_ctxt.clone(),
+                self,
+                TypeCollector {},
+                Linearization::new(HashMap::new()),
+            )
+            .map_err(|err| vec![Error::TypecheckError(err)])?;
+
+            lin_registry.insert(file_id, linearized, type_lookups, term);
```

Comment:
Why do we have to linearize twice and are we expecting to change that eventually? Could the generation of `type_lookups` be integrated into the implementation of `AnalysisHost` without a lot of effort?

Reply:
Yeah, good point. I would like to get rid of the first linearize pass eventually, but I think it might be a while so I've tidied this up in the meantime. I also got rid of a bunch of linearizer marker traits and wrappers that don't seem useful...